src/util/http/HttpClient.cpp

```diff
@@ -0,0 +1,116 @@
+// https://www.boost.org/doc/libs/master/libs/beast/example/http/client/sync-ssl/http_client_sync_ssl.cpp
+
+// ____________________________________________________________________________
+template <typename StreamType>
+void HttpClientImpl<StreamType>::openStream(const std::string& host,
+                                            const std::string& port) {
+  // IMPORTANT implementation note: Although we need only `stream_` later, it
+  // is important that we also keep `io_context_` and `ssl_context_` alive.
+  // Otherwise, we get a nasty and non-deterministic segmentation fault when
+  // the destructor of `stream_` is called. It seems that `resolver` does not
+  // have to stay alive.
+  if constexpr (std::is_same_v<StreamType, beast::tcp_stream>) {
+    tcp::resolver resolver{io_context_};
+    stream_ = std::make_unique<StreamType>(io_context_);
+    auto const results = resolver.resolve(host.c_str(), port.c_str());
+    stream_->connect(results);
+  } else {
+    static_assert(std::is_same_v<StreamType, ssl::stream<tcp::socket>>,
+                  "StreamType must be either boost::beast::tcp_stream or "
+                  "boost::asio::ssl::stream<boost::asio::ip::tcp::socket>");
+    ssl_context_ = std::make_unique<ssl::context>(ssl::context::sslv23_client);
+    ssl_context_->set_verify_mode(ssl::verify_none);
+    tcp::resolver resolver{io_context_};
+    stream_ = std::make_unique<StreamType>(io_context_, *ssl_context_);
+    if (!SSL_set_tlsext_host_name(stream_->native_handle(), host.c_str())) {
+      boost::system::error_code ec{static_cast<int>(::ERR_get_error()),
+                                   boost::asio::error::get_ssl_category()};
+      throw boost::system::system_error{ec};
+    }
+    auto const results = resolver.resolve(host.c_str(), port.c_str());
+    boost::asio::connect(stream_->next_layer(), results.begin(), results.end());
+    stream_->handshake(ssl::stream_base::client);
+  }
+}
+
+// ____________________________________________________________________________
+template <typename StreamType>
+std::istringstream HttpClientImpl<StreamType>::sendRequest(
+    const boost::beast::http::verb& method, const std::string& host,
+    const std::string& target, const std::string& requestBody,
+    const std::string& contentTypeHeader, const std::string& acceptHeader) {
+  // Check that we have a stream (obtained via a call to `openStream` above).
+  if (!stream_) {
+    throw std::runtime_error("Trying to send request without connection");
+  }
+
+  // Set up the request.
+  http::request<http::string_body> request;
+  request.method(method);
+  request.target(target);
+  request.set(http::field::host, host.c_str());
+  request.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+  request.set(http::field::accept, acceptHeader.c_str());
+  request.set(http::field::content_type, contentTypeHeader.c_str());
+  request.set(http::field::content_length, std::to_string(requestBody.size()));
```

Comment:
I'll also have to check:
Do these have to be `const char*` or can we also specify `string_view` here?
In that case the arguments to the function can also become `string_view` (which should be the default for function parameters wherever possible).